### ISSUE TITLE
[Bug] Route Repetition Slicing in build_route_split Causes Severe Train/Val Data Leakage

### ISSUE THREAD

**Issue by RustamSheoran** [Mohammedmt10/self-driving-model-in-simulation#4]
## Labels
`bug`, `high`, `dataset`, `evaluation`

## Affected Files
- `dataset.py` (lines 107–165)
- `dataset.py` (lines 82–88)

## Problem Summary
`build_route_split()` attempts to hold out entire routes for validation. However, it sorts route folders by `_parse_route_num` and slices the folder list by index `[:n_train]` and `[n_train:]`.

## Root Cause Code
```python
# dataset.py lines 136-163
route_folders = sorted(
    (p for p in os.listdir(scenario_dir) if p.startswith("Route") and os.path.isdir(os.path.join(scenario_dir, p))),
    key=_parse_route_num,
)

n_val = int(round(len(route_folders) * val_fraction))
n_train = len(route_folders) - n_val

for route in route_folders[:n_train]:
    train_list.extend(_build_route_pairs(os.path.join(scenario_dir, route)))
for route in route_folders[n_train:]:
    val_list.extend(_build_route_pairs(os.path.join(scenario_dir, route)))
```

## Why It Is Completely Ruining Everything
CARLA benchmark scenarios contain multiple repetitions of each route (e.g. `Route1_Rep0`, `Route1_Rep1`, `Route1_Rep2`). Because slicing cuts directly through the folder list, `Route1_Rep0` and `Route1_Rep1` end up in the training set, while `Route1_Rep2` ends up in the validation set. 

The validation set does not test unseen routes at all; it evaluates on the exact same road geometry and trajectory that the network trained on, yielding artificial validation scores that conceal severe overfitting.

## Proposed Fix
Extract unique route numbers first, partition route numbers into train/val subsets, and ensure all repetitions of a route number reside exclusively in one split.
